File: implementations/sson.cs
using System;
using static System.Console;
using System.Collections.Generic;

namespace sson
{
    public static class SSON
    {
        public static bool TrySetObjects(out Dictionary<string, Dictionary<string, string>> objects, string[] rawObjectData)
        {
            objects = new Dictionary<string, Dictionary<string, string>>();
            var defaultValues = new Dictionary<string, Dictionary<string, string>>();

            var readingObject = false; 
            var readingDefault = false;

            var lineCount = 0;
            var currentObject = "";

            foreach (var str in rawObjectData)
            {
                ++lineCount;

                if (str.StartsWith("#") || str.Length == 0) continue;

                if (str.StartsWith("."))
                {
                    if (!readingObject)
                    {
                        Error.WriteLine($"{str} at line {lineCount} is supposed to be a property, however it is cut off from its parent object, you probably misplaced a ; just before that line.");
                        return false;
                    }

                    var keyValuePair = str.Split('=');

                    if (keyValuePair.Length < 2)
                    {
                        Error.WriteLine($"expected a value after {str} at line {lineCount}; property cannot be empty");
                        return false;
                    }

                    // remove the dot at the beginning of the attribute
                    keyValuePair[0] = keyValuePair[0].Substring(1);

                    // removes the ; from the value if it's at the end of it
                    if (keyValuePair[1].EndsWith(";"))
                        keyValuePair[1] = keyValuePair[1].Remove(keyValuePair[1].Length - 1);

                    keyValuePair[0] = keyValuePair[0].Trim();
                    keyValuePair[1] = keyValuePair[1].Trim();

                    // put the value in the appropriate dictionary
                    if (readingDefault)
                    {
                        if (!defaultValues.ContainsKey(currentObject)) 
                            defaultValues.Add(currentObject, new Dictionary<string, string>());

                        if (!defaultValues[currentObject].ContainsKey(keyValuePair[0]))
                            defaultValues[currentObject].Add(keyValuePair[0], keyValuePair[1]);

                        else defaultValues[currentObject][keyValuePair[0]] = keyValuePair[1];
                    }

                    else
                    {
                        if (!objects.ContainsKey(currentObject))
                            objects.Add(currentObject, new Dictionary<string, string>());

                        if (!objects[currentObject].ContainsKey(keyValuePair[0]))
                            objects[currentObject].Add(keyValuePair[0], keyValuePair[1]);

                        else objects[currentObject][keyValuePair[0]] = keyValuePair[1];
                    }
                }

                else
                {
                    if (readingObject)
                    {
                        Error.WriteLine($"expect a ; before line {lineCount}");
                        return false;
                    }

                    if (str.StartsWith("default"))
                    {
                        readingObject = true;
                        readingDefault = true;

                        // removes the default part of the string.
                        currentObject = str.Substring(7).Trim();
                    }

                    else
                    {
                        var cleanStr = str.Trim();

                        currentObject = $"{cleanStr}_{lineCount}";
                        readingObject = true;

                        if (defaultValues.ContainsKey(cleanStr))
                        {
                            objects.Add(currentObject, new Dictionary<string, string>());

                            foreach (var property in defaultValues[cleanStr])
                                objects[currentObject].Add(property.Key, property.Value);
                        }
                    }
                }

                if (str.EndsWith(";"))
                {
                    if (!readingObject) 
                        Error.WriteLine($"redundant ; at line {lineCount}");

                    readingObject = false;
                    readingDefault = false;
                }
            }

            return true;
        }
    }
}

File: implementations/sson.vb
Imports System.Collections.Generic

Public Module SSON
	
	Function TrySetObjects(ByRef objects As Dictionary(Of String, Dictionary(Of String, String)), rawObjectData As String()) As Boolean
		
		objects = New Dictionary(Of String, Dictionary(Of String, String))()
		Dim defaultValues = New Dictionary(Of String, Dictionary(Of String, String))()
		
		Dim readingObject = False
		Dim readingDefault = False
		
		Dim lineCount = 0
		Dim currentObject = ""
		
		For Each str As String In rawObjectData
			
			lineCount += 1
			
			If str.StartsWith("#") Or str.Length = 0 Then
				
				Continue For
			End If
			
			If str.StartsWith(".") Then
				
				If Not readingObject Then
					
					Console.Error.WriteLine("{0} at line {1} is supposed to be a property, however it is cut off from its parent object, you probably misplaced a ; just before that line.", str, lineCount)
					Return False
				End If
				
				Dim keyValuePair = str.Split("=".ToCharArray(), 2)
				
				If keyValuePair.Length < 2 Then
					
					Console.Error.WriteLine("expected a value after {0} at line {1}; property cannot be empty", str, lineCount)
					Return False
				End If
				
				' remove the dot at the beginning of the attribute
				keyValuePair(0) = keyValuePair(0).Substring(1)
				
				' removes the ; from the value if it's at the end of it
				If keyValuePair(1).EndsWith(";") Then
					
					keyValuePair(1) = keyValuePair(1).Remove(keyValuePair(1).Length - 1)
				End If
				
				keyValuePair(0) = keyValuePair(0).Trim()
				keyValuePair(1) = keyValuePair(1).Trim()
				
				If readingDefault Then
					
					If Not defaultValues.ContainsKey(currentObject) Then
						
						defaultValues.Add(currentObject, New Dictionary(Of String, String)())
					End If
					
					If Not defaultValues(currentObject).ContainsKey(keyValuePair(0)) Then
						
						defaultValues(currentObject).Add(keyValuePair(0), keyValuePair(1))
						
					Else
						
						defaultValues(currentObject)(keyValuePair(0)) = keyValuePair(1)
					End If
					
				Else
					
					If Not objects.ContainsKey(currentObject) Then
						
						objects.Add(currentObject, New Dictionary(Of String, String)())
					End If
					
					If Not objects(currentObject).ContainsKey(keyValuePair(0)) Then
						
						objects(currentObject).Add(keyValuePair(0), keyValuePair(1))
						
					Else
						
						objects(currentObject)(keyValuePair(0)) = keyValuePair(1)
					End If
				End If
				
			Else If str <> ";" Then
				
				If readingObject Then
					
					Console.Error.WriteLine("expect a ; before line {0}", lineCount)
					Return False
				End If
				
				If str.StartsWith("default") Then
					
					readingObject = True
					readingDefault = True
					
					' removes the default part of the string.
					currentObject = str.Substring(7).Trim()
					
				Else
					
					Dim cleanStr = str.Trim()
					
					currentObject = String.Format("{0}_{1}", cleanStr, lineCount)
					readingObject = True
					
					If defaultValues.ContainsKey(cleanStr) Then
						
						objects.Add(currentObject, New Dictionary(Of String, String)())
						
						For Each attribute As KeyValuePair(Of String, String) In defaultValues(cleanStr)
							
							objects(currentObject).Add(attribute.Key, attribute.Value)
						Next
					End If
				End If
			End If
			
			If str.EndsWith(";") Then
				
				If Not readingObject Then
					
					Console.Error.WriteLine("redundant ; at line {0}", lineCount)
				End If
				
				readingObject = False
				readingDefault = False
			End If
		Next
		
		Return True
	End Function
End Module
